test/exercises/diffusion/heat_2d_dist_stencil.chpl

```diff
@@ -0,0 +1,74 @@
+/*
+  A distributed 2D finite-difference heat/diffusion equation solver
+
+  Computation is executed over a 2D distributed array.
+  The array distribution and sharing of "fluff"/"halo"
+  regions are managed by the `Stencil` distribution. The
+  `forall` loop manages task creation and synchronization
+  across and within locales.
+*/
+
+import StencilDist.Stencil,
+       Time.stopwatch;
+
+// create a stopwatch to time kernel execution
+var t = new stopwatch();
+config const writeTime = false;
+
+// compile with `-sRunCommDiag=true` to see comm diagnostics
+use CommDiagnostics;
+config param RunCommDiag = false;
+
+// declare configurable constants with default values
+config const nx = 256,      // number of grid points in x
+             ny = 256,      // number of grid points in y
+             nt = 50,       // number of time steps
+             alpha = 0.25,  // diffusion constant
+             solutionStd = 0.222751; // known solution for the default parameters
+
+// define a distributed 2D domain and subdomain to describe the grid and its interior
+const indices = {0..<nx, 0..<ny},
+      indicesInner = indices.expand(-1),
+      Indices = indices dmapped Stencil(indicesInner, fluff=(1,1)),
+      IndicesInner = Indices[{1..<nx-1, 1..<ny-1}];
+
+// define a distributed 2D array over the above domain
+var u: [Indices] real;
+
+// set up initial conditions
+u = 1.0;
+u[nx/4..nx/2, ny/4..ny/2] = 2.0;
+
+// start comm diag
+if RunCommDiag then startCommDiagnostics();
+
+// create a temporary copy of 'u' to store the previous time step
+var un = u;
+
+// iterate for 'nt' time steps
+t.start();
+for 1..nt {
+  // swap arrays to prepare for next time step
+  u <=> un;
+
+  // swap halo regions between neighboring tasks
+  un.updateFluff();
+
+  // compute the FD kernel in parallel
+  forall (i, j) in IndicesInner do
+    u[i, j] = un[i, j] + alpha *
+      (un[i-1, j-1] + un[i, j-1] + un[i+1, j] + un[i, j+1] - 4 * un[i, j]);
```

Comment:
Hmm, I could have sworn I fixed that. Let me check if it passes (or if I forgot to push some commits).

Reply:
Sorry about that, it looks like I forgot to push one last commit. This is fixed now.

Reply:
Note that one way such errors can be avoided in Chapel is to define tuples representing the four cardinal directions, like:

```chapel
const north = (-1, 0),
      south = (1, 0),
      east = (0, 1),
      west = (0,-1);
```

at which point, the loop can be written:

```chapel
forall ij in IndicesInner do
  u[ij] = un[ij] + alpha * (un[ij+north] + un[ij+south] + un[ij+east] + un[ij+west] - 4*un[ij]);
```

which is a verbosity / clarity tradeoff.  One downside to taking this approach in your code is that you've already got some variables that are effectively named north, south, east, west.

(another approach, which I don't recommend here, is to create an array or tuple of the offsets and to replace the four-expression add with a reduce like `+ reduce [off in offsets] un[ij+off]`.  This is a bit heavyweight (computationally) for a 5-point stencil, though, and a bit complex to teach someone as their first introduction to Chapel.  But for a larger stencil (27-point, 216-point), it's a big win in terms of not having to type out that many specific expressions).

For my own curiosity: Writing it out as I have here is the first time I've looked at the last term of the stencil.  Mathematically, isn't:

```chapel
un[ij] + alpha*(... - 4*un[ij])
```

equivalent to something like:

```chapel
(1 - alpha*4*un[ij] + alpha*(....))
```

(where I'm trying to optimize away the duplicate array indexing operation, since that can be expensive on a distributed array).  I suppose the downside is two multiplications by `alpha`?  Are there other reasons from a "true to the math/science" or "numerical stability" or "true to [someone else's version]" perspective to write it as you have?

Reply:
The current expression is based on the blog posts that I used to create earlier versions of this code. This is an artifact of the more complex expression used in those codes that includes separate constants for the X and Y portions of the stencil. That expression looks something like:

```chapel
u[i,j] = un[i, j] -
  alpha_x * (u[i-1, j] - 2*u[i, j] + u[i+1, j)) +
  alpha_y * (u[i, j-1] - 2*u[i, j] + u[i, j+1));
```

Would you expect the extra access to incur a performance penalty, or would you remove it more for the sake of improved simplicity in the code. I had originally assumed there would not be a performance penatly, under the assumption that the value would be available in the cache after the first access, so the second access would be very low cost.

There may also be numerical stability issues here — I'd need to investigate.

Reply:
OK, I'm definitely fine with keeping it as-is based on that rationale.  This is just a different 5-point stencil than I'm used to (a simple Jacobi iteration).

I would expect it to incur some performance penalty today because we're not good at common subexpression elimination (CSE) in which the compiler might rewrite the original as:

```chapel
const unij = un[i, j];  // or `const ref unij = ...;`
u[i,j] = unij - alpha*(... + 4*unij);
```

to avoid the double array reference.  Block array references are expensive today because they basically have to say "Is this local or remote; if local, I'll just access the element and return it; if remote, then I'll need to communicate to a remote locale to get the value".  It's this complexity that also (almost certainly) prevents the back-end C/LLVM compiler from doing the CSE itself—lots of side effects and calls to the runtime that will thwart a compiler.

Since the value will be local, the good thing is we're not communicating twice, but we will still do a bunch of control flow and indexing to get that local element twice I think.  You could do a manual rewrite like the above to see how big the impact is.

I think this discussion is mostly academic, though...  it probably will hurt the performance a bit, but I wouldn't clutter up the code because of it unless the gains were massive.  And it feels unfair to optimize the block/stencil versions in this way, but not the SPMD version (and the SPMD version is already plenty complex without this sort of manual optimization).

This is also what spawned the suggestion about (ultimately, not necessarily today) having a block + local access version.  That would bring the overheads for the duplicate array reference down to be similar to the SPMD version (where there are still overheads due to the double-deref, but lower ones since the arrays are known to be local).

Now, if we were racing against a C+MPI version, that might be a reason to consider such optimizations...  but not for a tutorial, IMO.